src/utils/github_updater.rs

```diff
@@ -16,115 +24,153 @@ struct GitHubFields {
+            (ACCEPT, HeaderValue::from_static("application/json")),
+        ];
+
+        if let Some((username, accesstoken)) = config.github_auth() {
+            let basicauth = format!(
+                "Basic {}",
+                base64::encode(format!("{}:{}", username, accesstoken))
+            );
+            headers.push((AUTHORIZATION, HeaderValue::from_str(&basicauth).unwrap()));
+        } else {
+            warn!("No GitHub authorization specified, will be working with very low rate limits");
         }
 
-        // sleep for rate limits
-        use std::thread;
-        use std::time::Duration;
-        thread::sleep(Duration::from_secs(2));
+        let client = reqwest::blocking::Client::builder()
+            .default_headers(headers.into_iter().collect())
+            .build()?;
+
+        Ok(GithubUpdater { client, pool })
     }
 
-    Ok(())
-}
+    /// Updates github fields in crates table
+    pub fn update_all_crates(&self) -> Result<()> {
+        debug!("Starting update of all crates");
+
+        if self.is_rate_limited()? {
+            warn!("Skipping update because of rate limit");
+            return Ok(());
+        }
+
+        let conn = self.pool.get()?;
+        // TODO: This query assumes repository field in Cargo.toml is
+        //       always the same across all versions of a crate
+        let rows = conn.query(
+            "SELECT DISTINCT ON (crates.name)
+                    crates.name,
+                    crates.id,
+                    releases.repository_url
+             FROM crates
+             INNER JOIN releases ON releases.crate_id = crates.id
+             WHERE releases.repository_url ~ '^https?://github.com' AND
+                   (crates.github_last_update < NOW() - INTERVAL '1 day' OR
+                    crates.github_last_update IS NULL)
+             ORDER BY crates.name, releases.release_time DESC",
+            &[],
+        )?;
+
+        for row in &rows {
+            let crate_name: String = row.get(0);
+            let crate_id: i32 = row.get(1);
+            let repository_url: String = row.get(2);
+
+            debug!("Updating {}", crate_name);
+            if let Err(err) = self.update_crate(&conn, crate_id, &repository_url) {
+                if self.is_rate_limited()? {
+                    warn!("Skipping remaining updates because of rate limit");
+                    return Ok(());
+                }
+                warn!("Failed to update {}: {}", crate_name, err);
+            }
+        }
+
+        debug!("Completed all updates");
+        Ok(())
+    }
+
+    fn is_rate_limited(&self) -> Result<bool> {
+        #[derive(Deserialize)]
+        struct Response {
+            resources: Resources,
+        }
+
+        #[derive(Deserialize)]
+        struct Resources {
+            core: Resource,
+        }
+
+        #[derive(Deserialize)]
+        struct Resource {
+            remaining: u64,
+        }
```

Comment:
This seems more complicated than using a json library directly?

Reply:
It's faster, type safe, more extensible and more maintainable so I prefer this option to another [`Metadata::from_str`](https://github.com/rust-lang/docs.rs/blob/master/src/docbuilder/metadata.rs#L129-L179)

Reply:
The alternative would be ~

```rust
response
  .as_object()
  .context("invalid type, expected object")?
  .get("resources")
  .context("missing field `resources`")?
  .as_object()
  .context("invalid type, expected object")?
  .get("core")
  .context("missing field `core`")?
  .as_object()
  .context("invalid type, expected object")?
  .get("remaining")
  .context("missing field `remaining`")?
  .as_u64()
  .context("invalid type, expected u64")?
```

though even that is missing some of the error context, e.g. `serde` will give the actual type when it's wrong; and it's using `anyhow`'s more capable `context` while with `failure` those would have to be `.ok_or_else(|| err_msg("invalid type, expected u64"))?`.

So it actually ends up being the exact same number of lines of code, and I find the declarative data structure much easier to read.

<details><summary>Maybe in 2042 we'll be able to write this nicely with anonymous structs</summary>

```rust
#[derive(Deserialize)]
struct Response {
    resources: {
        core: {
            remaining: u64,
        },
    },
}
```

</details>

Reply:
Ouch, yeah I didn't realize the alternative was so wordy. This looks fine then.